go-controller/pkg/ovn/logical_switch_manager.go

```diff
@@ -0,0 +1,255 @@
+	// make a deep-copy of the underlying slice and return so that there is no
+	// resource contention
+	if ok && len(lsi.hostSubnets) > 0 {
+		subnets := make([]*net.IPNet, len(lsi.hostSubnets))
+		for i, hsn := range lsi.hostSubnets {
+			subnet := *hsn
+			subnets[i] = &subnet
+		}
+		return subnets
+	}
+	return nil
+}
+
+// AllocateIPs will block off IPs in the ipnets slice as already allocated
+// for a given switch
+func (manager *logicalSwitchManager) AllocateIPs(nodeName string, ipnets []*net.IPNet) error {
+	manager.RLock()
+	defer manager.RUnlock()
+	lsi, ok := manager.cache[nodeName]
+	if len(ipnets) <= 0 || !ok || (ok && len(lsi.ipams) <= 0) {
+		return fmt.Errorf("unable to allocate ips %v for node: %s",
+			ipnets, nodeName)
+
+	}
+	for _, ipnet := range ipnets {
+		for _, ipam := range lsi.ipams {
+			cidr := ipam.CIDR()
+			if cidr.Contains(ipnet.IP) {
+				if err := ipam.Allocate(ipnet.IP); err != nil {
+					return err
+				}
+				break
+			}
+		}
+	}
+	return nil
+}
+
+// AllocateNextIPs allocates IP addresses from each of the host subnets
+// for a given switch
+func (manager *logicalSwitchManager) AllocateNextIPs(nodeName string) ([]*net.IPNet, error) {
+	manager.RLock()
+	defer manager.RUnlock()
+	var ipnets []*net.IPNet
+	var ip net.IP
+	var err error
+	lsi, ok := manager.cache[nodeName]
+	if ok && len(lsi.ipams) == 0 {
+		return nil, fmt.Errorf("failed to allocate IPs for node %s because there is no IPAM instance", nodeName)
+	}
+
+	if len(lsi.ipams) != len(lsi.hostSubnets) {
+		return nil, fmt.Errorf("failed to allocate IPs for node %s because host subnet instances: %d"+
+			" don't match ipam instances: %d", nodeName, len(lsi.hostSubnets), len(lsi.ipams))
+	}
+
+	defer func() {
+		if err != nil {
+			// iterate over range of already allocated indices and release
+			// ips allocated before the error occurred.
+			for relIdx, relIPNet := range ipnets {
+				if relErr := lsi.ipams[relIdx].Release(relIPNet.IP); relErr != nil {
+					klog.Errorf("Error while releasing IP: %s, err: %v", relIPNet.IP, relErr)
+				}
+			}
+			klog.Warningf("Allocated IPs: %s were released", util.JoinIPNetIPs(ipnets, " "))
+		}
+	}()
+
+	for idx, ipam := range lsi.ipams {
+		ip, err = ipam.AllocateNext()
+		if err != nil {
+			return nil, err
+		}
+		ipnet := &net.IPNet{
+			IP:   ip,
+			Mask: lsi.hostSubnets[idx].Mask,
+		}
+		ipnets = append(ipnets, ipnet)
+	}
+	return ipnets, nil
+}
+
+// Mark the IPs in ipnets slice as available for allocation
+// by releasing them from the IPAM pool of allocated IPs.
+func (manager *logicalSwitchManager) ReleaseIPs(nodeName string, ipnets []*net.IPNet) error {
+	manager.RLock()
+	defer manager.RUnlock()
+	if ipnets == nil || nodeName == "" {
+		klog.V(5).Infof("Node name is empty or ip slice to release is nil")
+		return nil
+	}
+	lsi, ok := manager.cache[nodeName]
+
+	if ok && len(lsi.ipams) == 0 {
+		return fmt.Errorf("failed to release IPs for node %s because there is no IPAM instance", nodeName)
+	}
```

Comment:
What if `ok` is false here and `lsi` will be nil and at L244, `lsi.ipams` will panic, right?

Reply:
done